### ISSUE TITLE
Design easy install prompts

### ISSUE THREAD

**Issue by bartkamphorst** [repotag/repotag#14]
Ideally, Repotag starts with a few prompts for:
- [ ] Creating an admin user
- [ ] ~~Dumping a configured sqlite database on disk~~ Configuring database access
- [ ] Setting default paths for repos, wikis and archives
- [ ] Create and save production secret key

**Comment by bartkamphorst** [comment]
> Create and save production secret key

See http://www.jamesbadger.ca/2012/12/18/generate-new-secret-token/

**Comment by dometto** [comment]
There is also the question of database configuration. As I think you've noted in our current `database.yml`, we can make the db configurable through command line options by warbling a `database.yml` like this one:

``` erb
production:
  <% if java.lang.System.getProperty('db_adapter') %>
  adapter: <%= java.lang.System.getProperty('db_adapter') %>
  database: <%= (java.lang.System.getProperty('db_name') %>
  host: <%= java.lang.System.getProperty('db') %>
  username: <%= java.lang.System.getProperty('username') %>
  password: <%= java.lang.System.getProperty('password') %>
  <% else %>
  adapter: sqlite3
  database: <%= java.lang.System.getProperty('db_name') || 'mydb.sqlite3' %>
  <% end %>
```

Another option would be to use `ENV['DATABASE_NAME']`, etc. The advantage of that would perhaps be that you don't need to pass the db password as a command line argument (which would show up in process names and such).

**Comment by bartkamphorst** [comment]
Yes, but ideally, the user could configure all this at the first run. :smile:

**Comment by dometto** [comment]
That would be awesome, but not sure if we could pull it off. Rails (or at least ActiveRecord) just won't load without a db connection I think?

Maybe we could put a middleware before rails to take care of first-run config? But then we'd have to figure out how to warble that and start rails from it.

**Comment by bartkamphorst** [comment]
> Maybe we could put a middleware before rails to take care of first-run config? But then we'd have to figure out how to warble that and start rails from it.

Yes, I had something in mind like that.

**Comment by bartkamphorst** [comment]
There are (at least) two ways to go, it seems. Currently, we are building an executable war with warbler, which means it uses `config.ru` to start Rails. We could build a wrapper class that starts one app (e.g., Sinatra) on first boot and Rails otherwise. The second way is to trick warbler into building a jar instead of a war, in which case the first script in `bin` should run. In that case, we could distinguish between the setup app and the regular app there. As it stands, I favor the first option. You?

**Comment by bartkamphorst** [comment]
> We could build a wrapper class that starts one app (e.g., Sinatra) on first boot and Rails otherwise.

We should check if we have access at that point to commandline options via `java.lang.System.getProperty('setup')` for example.

**Comment by dometto** [comment]
Yes, I also favour the first. Maybe it wouldn't even have to be a Sinatra app but just a Rack::Builder object serving static pages.

**Comment by bartkamphorst** [comment]
Could well be. All we need in the beginning is one form that triggers a number of configuration actions.

**Comment by bartkamphorst** [comment]
Another alternative could be to make the war _runnable_, which would allow us to run binaries such as rake (like so `java -jar myrailsapp.war -S rake db:migrate`).

**Comment by dometto** [comment]
Sinatra might be nice though, if we want to use bootstrap as a layout for the first-run form (because then we need use do `less` or `sass`). So maybe we have:
1. a simple `#call`able middleware object that checks whether all the required configuration options are set
2. if yes, it loads Rails
3. if no, it loads the sinatra first-run app

That way we don't suffer the performance penalty of loading sinatra before every rails action. Some worries:
- do we need to do the `dup._call` trick for the first middleware for thread safety?
- can we use the bootstrap assets from repotag in the sinatra app?
- if we want to be able to configure an admin user from the sinatra app, it would have to be able to setup the repotag database schema. Not sure if it's possible to do that cleanly?

**Comment by dometto** [comment]
> Another alternative could be to make the war runnable, which would allow us to run binaries such as rake (like so java -jar myrailsapp.war -S rake db:migrate).

And then require the user to do call a first-run command from the CLI if no config options are found? That would solve the problem of how to setup the database.

**Comment by dometto** [comment]
> Sinatra might be nice though, if we want to use bootstrap as a layout for the first-run form (because then we need use do less or sass)

Ah, no -- in production, the assets will be compiled, so we could serve them statically. Sinatra might still be required for database operations.

**Comment by bartkamphorst** [comment]
It depends on what we want with the database. Warbler bundles `production.sqlite3` so in principle we could just copy that to a user-specified directory and reference that path in `.repotag.yml` or something. If we want to enable mysql or postgres, however, we will need to be able to run a rake task or something to load the db schema.

**Comment by dometto** [comment]
>  If we want to enable mysql or postgres, however, we will need to be able to run a rake task or something to load the db schema.

Right. I think we need to support mysql and postgres though. Is there a way to run such a rake task from inside a sinatra/rack app?

**Comment by bartkamphorst** [comment]
It's Ruby. :smile: You can do something like this

``` ruby
Rake::Task['task_name'].invoke
```

to run a rake task programmatically.

**Comment by dometto** [comment]
Cool. Also see http://stackoverflow.com/a/26046277

**Comment by dometto** [comment]
Also https://github.com/janko-m/sinatra-activerecord

**Comment by dometto** [comment]
I got rake db tasks working outside of the rails app. See https://gist.github.com/dometto/db58186b37c1e7ff9f99

Run `test.rb` with your repotag gemset and it will add the admin user to your development database (this is because `test.rb` sets the rails env to `development`)

**Comment by dometto** [comment]
With [this](https://gist.github.com/dometto/db58186b37c1e7ff9f99#file-config-ru) `config.ru` I can succesfully run _either_ Repotag _or_ a Sinatra app depending on a command line option (`setup_db`). Place the `config.ru` in repotag's root (replacing the current file).

With this you get repotag:
`ruby -J-Dsetup_db=true /path/to/rackup config.ru`

With this you get the Sinatra app's output:
`ruby /path/to/rackup config.ru`

However, if I run `warble` in the repotag directory, it just packages the rails app, ignoring `config.ru`. I'm not sure if we can force warbler to treat the app as a rack app rather than as a Rails app. But maybe we can somehow work around this issue.

Anyway, we should decide when exactly to run the config/firstrun app. When there is no admin user? When certain settings aren't found in the db?

Perhaps for db configuration it would be better to just have `config=path/to/repotag.yml`? Otherwise you would have to reconfigure the app everytime you restart it. We will still need to offer a way to initialize a new database, but maybe it's better to do that by making the war runnable than through the firstrun app?

**Comment by dometto** [comment]
> Perhaps for db configuration it would be better to just have config=path/to/repotag.yml? Otherwise you would have to reconfigure the app everytime you restart it.

The more I think about this the more I feel a separate config/firstrun app is a nonstarter... there is no such thing as a 'first run' if we don't store the config somewhere (e.g. in a .yml).

**Comment by bartkamphorst** [comment]
All good points. Here's roughly what I have in mind.
- User runs `java -jar repotag.war` or `java -jar repotag.war -Dconfig=/path/to/repotag.yml`. In the first case, custom Rack middleware looks for the config file in cwd.
- If no config is found, the app errors out, suggesting to point to the config or to run `java -jar repotag.war -Dsetup`.
- Running repotag with the setup parameter starts a basic Sinatra app in which users can specify how they want their database set up (and perhaps some other things). This info is then written to `repotag.yml` in cwd. This app also performs some initialization functions if necessary/desired (test db connection, load schema, etc.).
- If the config _is_ found, the Rack middleware does a quick sanity check (is db available?) and makes these settings available to Rails through a set of environment variables, or, alternatively, it sets `ENV['repotag_config']` which Rails can then read in on initialization.

**Comment by dometto** [comment]
Sounds good. But if no config is found, do we want to error out or 
default to the sqlite db bundled in the jar? Alternatively we could do 
that only when the user explicitly sets an `sqlite` parameter, but maybe 
that's too complicated.

**Comment by bartkamphorst** [comment]
If we default to the bundled sqlite3 db then we have to make clear to the user that changes will only persist until the server is terminated (so that it is basically a demo run).

**Comment by dometto** [comment]
Regarding starting the Sinatra app, I’m not sure exactly how warbler 
calls rails. If it does so by requiring |config/environment.rb|, maybe 
we can just hack that file and do:

|java.lang.System.getProperty('setup') ? Sinatra.foo : 
Rails::Application.initialize! |

​

**Comment by dometto** [comment]
Ah, I somehow assumed that the in-jar db could be modified. Actually it 
makes it simpler if we don't have to consider that option, so good. :-)

**Comment by bartkamphorst** [comment]
According to [the docs](https://github.com/jruby/warbler#war-or-jar) it should use either `config/environment.rb` or `config.ru`.

**Comment by bartkamphorst** [comment]
Minimally, this is what `repotag.yml` should have I think (secret is randomly generated and for demonstration purposes only).

``` yaml
---
:database:
  :adapter: mysql
  :encoding: utf8
  :database: repotag
  :username: root
  :password: mypw
  :host: localhost
  :port: 3306
:secret_key_base: 5d6537d631df9555f6f7071ca60e91f1c633a49851dd922d597982de998b0fe7ee828c1e82f042567c03b0c2b24228f9442a6f928a6ed757c7bbd32c201f4667
```
